checkArticleExists throwing on error

**Describe the bug**
For unclear reasons, occassionally checkArticleExists throws with an `MongoNetworkError: write EPIPE` error. Whilst this deserves its own investigation, the bigger issue is that this is a show-stopping throw rather than an error that is merely reported to log.

**To Reproduce**
No sure, presumably this is coming from the feed-checking function.

**Expected behavior**
When error occurs in `checkArticleExists` it is logged to console and app continues on.

**Error message**
```
/usr/src/app/node_modules/mongodb/lib/topologies/server.js:240
            throw err;
            ^

AssertionError [ERR_ASSERTION]: null === 'MongoNetworkError: write EPIPE'
    at /usr/src/app/lib/queries.js:337:14
    at err (/usr/src/app/node_modules/mongodb/lib/utils.js:411:14)
    at executeCallback (/usr/src/app/node_modules/mongodb/lib/utils.js:401:25)
    at err (/usr/src/app/node_modules/mongodb/lib/operations/mongo_client_ops.js:285:21)
    at connectCallback (/usr/src/app/node_modules/mongodb/lib/operations/mongo_client_ops.js:265:5)
    at server.connect (/usr/src/app/node_modules/mongodb/lib/operations/mongo_client_ops.js:353:14)
    at Server.<anonymous> (/usr/src/app/node_modules/mongodb/lib/topologies/server.js:237:11)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at Server.emit (events.js:211:7)
    at Pool.<anonymous> (/usr/src/app/node_modules/mongodb-core/lib/topologies/server.js:452:12)
    at emitOne (events.js:116:13)
    at Pool.emit (events.js:211:7)
    at connect (/usr/src/app/node_modules/mongodb-core/lib/connection/pool.js:557:14)
    at callback (/usr/src/app/node_modules/mongodb-core/lib/connection/connect.js:109:5)
    at runCommand (/usr/src/app/node_modules/mongodb-core/lib/connection/connect.js:129:7)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! rockpool-app@1.0.0 start: `NODE_ENV=production node app.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the rockpool-app@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-07-05T04_06_32_552Z-debug.log
```

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04

**Smartphone (please complete the following information):**
n/a

**Additional context**
n/a

--- a/lib/feeds.js
+++ b/lib/feeds.js
@@ -20,6 +20,7 @@ const queries = require('./queries.js') // local database queries module
 const blogs = require('./blogs.js') // local database blogs module
 const announcements = require('./announcements.js') // local database blogs module
 const { unsubscribeFromPocket  } = require('./users.js') // local database updates module
+const { sendEmail } = require('./utilities')
 // *********
 // FUNCTIONS
 // *********
@@ -71,6 +72,52 @@ function addArticle(blog, item) {
   })
 }
 
+function sendArticleToPocket(user, url) {
+  return axios.post('https://getpocket.com/v3/add', {
+    consumer_key: settings.pocket_consumer_key,
+    access_token: user.pocket.token,
+    url: url,
+    tags: settings.app_name
+  })
+  .then( () => {
+    resolve()
+  })
+  .catch(err => {
+    // catch errors and report them
+    // this is likely to be either something wrong with the pocket server
+    // or maybe a token has been removed but not deleted in the app
+    console.error(`error when processing pocket account for ${user.email}`)
+    if (err.response) {
+      // error response, most likely a 401
+      console.log(err.response.status + ' - ' + err.response.headers['x-error'])
+      if (err.response.status === 401) {
+        // 401 is 'unauthorised' meaning the pocket access token
+        // isn't recognised (we can assume it's been revoked)
+        // so here we remove the user's pocket credentials
+        // then email them to let them know
+        return unsubscribeFromPocket(user.email)
+        .then( () => {
+          let message = {
+            to: user.email,
+            subject: `Your pocket account has been removed from ${settings.app_name}`,
+            text: `Your pocket account has been removed from ${settings.app_name} because the access token is no longer recognised. If you did not mean to revoke access, you can log in again at ${settings.app_url} to reconnect ${settings.app_name} to your Pocket account.`
+          }
+          return sendEmail(message)
+        })
+      }
+    } else if (error.request) {
+      // The request was made but no response was received
+      console.log('No response')
+      console.log(error.request)
+      return
+    } else {
+      // Something happened in setting up the request that triggered an Error
+      console.log('Error', error.message)
+      return
+    }
+  })
+}
+
 const pushToPockets = function(data) {
   return new Promise(function (resolve, reject) {
     if (data) {
@@ -86,53 +133,17 @@ const pushToPockets = function(data) {
           for (let user of res.users) {
             // check for pocket exclusions
             let pocketExcluded = user.pocket.excluded ? user.pocket.excluded.some( x => x.equals(data.id)) : false
-            function sendArticleToPocket() {
-              return axios.post('https://getpocket.com/v3/add', {
-                consumer_key: settings.pocket_consumer_key,
-                access_token: user.pocket.token,
-                url: data.url,
-                tags: settings.app_name
-              })
-              .then( () => {
-                resolve()
-              })
-              .catch(err => {
-                // catch errors and report them
-                // this is likely to be either something wrong with the pocket server
-                // or maybe a token has been removed but not deleted in the app
-                console.error(`error when processing pocket account for ${user.email}`)
-                if (err.response) {
-                  // error response, most likely a 401
-                  console.log(err.response.status + ' - ' + err.response.headers['x-error'])
-                  if (err.response.status === 401) {
-                    // 401 is 'unauthorised' meaning the pocket access token
-                    // isn't recognised (we can assume it's been revoked)
-                    // so here we remove the user's pocket credentials
-                    // TODO: ideally we'd communicate this to the user - via email?
-                    unsubscribeFromPocket(user.email)
-                    .then( () => {
-                      resolve()
-                    })
-                  }
-                } else if (error.request) {
-                  // The request was made but no response was received
-                  console.log('No response')
-                  console.log(error.request)
-                  resolve()
-                } else {
-                  // Something happened in setting up the request that triggered an Error
-                  console.log('Error', error.message)
-                  resolve()
-                }
-              })
-            }
+            // send to pocket if not excluded for this user
             if (!pocketExcluded) {
-              setTimeout(sendArticleToPocket, 2000) // delay each API call by 2 seconds
+              setTimeout(sendArticleToPocket(user, data.url), 2000) // delay each API call by 2 seconds
             }
           }
         } else {
           resolve()
         }
+      })
+      .then( () => {
+        resolve()
       })
       .catch(err => {
         reject(err)
@@ -153,8 +164,10 @@ const checkFeeds = function() {
     }) // get all feeds that are approved and aren't suspended
     .then( res => {
       for (let blog of res.blogs) {
-        return feedparser
-        .parse({uri: blog.feed, timeout: 10000}) // include a timeout to avoid hanging
+        return feedparser.parse({
+          uri: blog.feed, 
+          timeout: 10000
+        }) // include a timeout to avoid hanging
         .then( items => {
           return blogs.setFailing({
             url: blog.url,

--- a/lib/feeds.js
+++ b/lib/feeds.js
@@ -85,7 +85,7 @@ const pushToPockets = function(data) {
           // for each user
           for (let user of res.users) {
             // check for pocket exclusions
-            let excluded = user.pocket.excluded ? user.pocket.excluded.some( x => x.equals(data.id)) : false
+            let pocketExcluded = user.pocket.excluded ? user.pocket.excluded.some( x => x.equals(data.id)) : false
             function sendArticleToPocket() {
               return axios.post('https://getpocket.com/v3/add', {
                 consumer_key: settings.pocket_consumer_key,
@@ -126,7 +126,7 @@ const pushToPockets = function(data) {
                 }
               })
             }
-            if (!excluded) {
+            if (!pocketExcluded) {
               setTimeout(sendArticleToPocket, 2000) // delay each API call by 2 seconds
             }
           }
@@ -145,7 +145,7 @@ const pushToPockets = function(data) {
 
 const checkFeeds = function() {
   return new Promise(function (resolve, reject) {
-    queries.getBlogs({
+    return queries.getBlogs({
       query: {
         approved: true, // blogs must be approved
         suspended: {$ne: true} // blogs must not be suspended
@@ -153,16 +153,16 @@ const checkFeeds = function() {
     }) // get all feeds that are approved and aren't suspended
     .then( res => {
       for (let blog of res.blogs) {
-        feedparser
-        .parse({uri: blog.feed, timeout: 10000}) // include a timeout, otherwise it can hang forever
+        return feedparser
+        .parse({uri: blog.feed, timeout: 10000}) // include a timeout to avoid hanging
         .then( items => {
-          blogs.setFailing({
+          return blogs.setFailing({
             url: blog.url,
             failing: false
-          }) // set failing to false in case it was previously set to true
+          }) // ensure failing set to false
           .then( () => {
             if (items) {
-              items.forEach( function (item) {
+              return items.forEach( item => {
                 let downcased = item.categories.map( tag => tag.toLowerCase())
                 // check for exclude tags
                 if (downcased.some(excluded)) {
@@ -172,9 +172,7 @@ const checkFeeds = function() {
                 } else if ( settings.included_tags.length > 0 && !downcased.some(included) ) {
                   resolve(true) // if included_tags is not empty but no include tags are used in the post, ignore it.
                 } else {                  
-                // check whether is in the DB already - if it isn't, notInDB is false, else is true
-                //  if EITHER the guid or post URL match, notInDB is false
-                queries.checkArticleExists({
+                return queries.checkArticleExists({
                   item: item,
                   query: {
                     $or: [ 
@@ -185,13 +183,14 @@ const checkFeeds = function() {
                 })
                 .then( item => {
                   if (item) {
-                    addArticle(blog, item) // ingest into DB
+                    return addArticle(blog, item) // ingest into DB
                     .then( res => {
                       // check it is within the cutoff period for recency
                       let now = new Date()
                       let cutoff = new Date(now - 1.728e+8) // 48 hours ago
                       if (res.date > cutoff) {
-                        announcements.queueArticleAnnouncement(blog, res) // queue announcement if more recent than 48 hours
+                        // queue announcement if more recent than 48 hours
+                        return announcements.queueArticleAnnouncement(blog, res)
                         .then(pushToPockets)
                         .then( () => {
                           resolve(true)
@@ -210,7 +209,7 @@ const checkFeeds = function() {
             }
           })
         })
-        .catch(e => {
+        .catch( e => {
           blogs.setFailing({
             url: blog.url,
             failing: true

